### ISSUE TITLE
Author URLs don't need to be prefixed, do they?

### ISSUE THREAD

**Issue by aryamccarthy** [acl-org/acl-anthology#211]
[minor]

If I want to visit the anthology's author page for Matt Post, the URL is <http://aclweb.org/anthology/people/m/matt-post/>. Similarly, if I look at the author page for Dan Klein, the URL is <http://aclweb.org/anthology/people/d/dan-klein/>. Do we need these prefixes, like `/m/` and `/d/`? Or can the URLs be altered to use, say, <http://aclweb.org/anthology/people/matt-post>?

**Comment by mbollmann** [comment]
Without these prefixes, we end up with 200k+ files in a single directory, which (at least on my system) already slows down basic commands like `ls`. I don't know if it'd also affect web server performance. But it's mainly a pragmatic choice, not an aesthetic or idealistic one.

**Comment by aryamccarthy** [comment]
Ah, that's fair. If we want to also work the aesthetics in, we can use the `permalink:` attribute in each page's front matter. A nice long-term direction would be to generate YAML for the author's page instead. (I know YAML-ifying everything is a goal of yours.) After that, you could specify a permalink pattern for the entire collection, with or without the prefix character.

**Comment by mbollmann** [comment]
I'm not sure I follow. The author pages are already generated based on a YAML file; what usage do you envision for it?

If you mean aliasing `people/matt-post` to `people/m/matt-post` in the static site generator, that would be possible, but still cause the abovementioned problem. What *should* be doable is to add an Apache rewrite rule to do this, as this is how the PDF URLs already work (http://aclweb.org/anthology/P05-1021 links to http://aclweb.org/anthology/P/P05/P05-1021.pdf etc.).

**Comment by mjpost** [comment]
I added a Rewrite Rule:

- http://aclweb.org/anthology/people/dan-klein
- http://aclweb.org/anthology/people/arya-mccarthy

**Comment by mbollmann** [comment]
Cool! I'll check if I can make it work in Hugo to produce links with this scheme, too.

**Comment by mjpost** [comment]
Do we need to change them? I'm not arguing against doing so, or for it. I am ambivalent—so do what you like.

**Comment by mbollmann** [comment]
I don't have a strong feeling either, but we should decide on what should be the canonical variant.

+ If it's `people/dan-klein`, then the links on the website should point to that.
+ If it's `people/d/dan-klein`, the server should explicitly redirect the other variant there, and not alias it (as it seems to be now).

**Comment by mjpost** [comment]
Okay, if you're willing, let's do the shorter variant (`people/dan-klein`). This is more consistent with how we handle paper IDs, and also lets us reorganize more easily later on.

What letters are permitted in these URLs? I am currently using this rule:

    RewriteRule ^people/([a-z])([a-z\-]*)$ people/$1/$1$2/ [L,NC]

(Incidentally, what do you think about checking our `.htaccess` file into the repo?)

**Comment by mbollmann** [comment]
There are some caveats with this solution though: To produce links to pages that are different from what's actually on the filesystem, I might have to insert them directly and bypass linking to the actual page `.md` file, which 

1. makes Hugo unable to check if these links actually work (and produce an error message if not), and
2. makes it impossible for the site to work correctly on a mirror or locally without this rewrite rule. 

This could maybe be addressed by setting a "production" flag during generation that changes this behaviour, but it would be a bit more involved than what we currently have.

As to the allowed letters, it should probably be everything that can be output by [python-slugify](https://github.com/un33k/python-slugify), which I believe is `[-a-z0-9_]`.

**Comment by mjpost** [comment]
Yes, we do have our own. I'll attach it here since I'm not sure where to put it—can you add it?

Here is the command I use for rsyncing to the Anthology after a build, in case it's useful:

   rsync -azve ssh hugo/public/ hugo/data-export/ aclweb:/path/to/anthology

(Note: the `.htaccess` probably needs some cleanup).

```
# This is important for returning the proper encoding for BibTeX files
AddDefaultCharset utf-8

# Only runs for files or directories that are not found
RewriteEngine On
# RewriteBase /anthology/

# Normal papers (nothing at the end: the canonical URL)
#RewriteRule ^(E)([0-9][0-9])\-([0-9][0-9][0-9][0-9])$ /anthology-files/pdf/$1/$1$2/$1$2-$3.pdf [L,NC]
RewriteRule ^([A-Za-z])([0-9][0-9])\-([0-9][0-9][0-9][0-9])$ $1/$1$2/$1$2-$3.pdf [L,NC]
# PDF link
RewriteRule ^([A-Za-z])([0-9][0-9])\-([0-9][0-9][0-9][0-9]).pdf$ $1/$1$2/$1$2-$3.pdf [L,NC]
# Added by Min (18 Aug 2014) to cater for revisions/errata
RewriteRule ^([A-Za-z])([0-9][0-9])\-([0-9][0-9][0-9][0-9])([ve][0-9]+)$ $1/$1$2/$1$2-$3$4.pdf [L,NC]
# Added for attachments
RewriteRule ^([A-Za-z])([0-9][0-9])\-([0-9][0-9][0-9][0-9])(\..*)?$ $1/$1$2/$1$2-$3$4 [L,NC]
# Added for volumes (by Min, 25 Sep 2014)
RewriteRule ^([A-Za-z])([0-9][0-9])\-([0-9]{1,2})$ $1/$1$2/$1$2-$3.pdf [L,NC]

# Paper metadata pages (put / at end)
RewriteRule ^([A-Za-z])([0-9][0-9])\-([0-9][0-9][0-9][0-9])\/$ papers/$1/$1$2/$1$2-$3/ [L,NC]
# BibTeX
RewriteRule ^([A-Za-z])([0-9][0-9])\-([0-9][0-9][0-9][0-9])\.([a-z]+)$ papers/$1/$1$2/$1$2-$3.$4 [L,NC]

# Author pages
RewriteRule ^people/([a-z])([a-z\-]*)$ people/$1/$1$2/ [L,NC]


# ----------------------------------------------------------------------
# Ugly section for lowercase paths (to transform to uppercase) (by Min, 25 Sep 2014)

# # If there are caps, set HASLOWER to true and skip next rule
# RewriteRule [w] - [E=HASLOWER:TRUE,S=1]
 
# # Skip this entire section if no lowercase letters in requested URL
# RewriteRule ![w] - [S=3]
 
# # Replace single occurence of cap with CAP, then process next rule.
# RewriteRule ^([^w]*)w(.*)$ //$1W//$2//
 
# # If there are any uppercase letters, restart at very first RewriteRule in file.
# RewriteRule [w] - [N]
 
# RewriteCond %{ENV:HASLOWER} TRUE
# RewriteRule ^/?(.*) /$1 [R=301,L]
# # ----------------------------------------------------------------------
# Old rules (attachments for revisions/errata?)
# RewriteRule ^([A-Za-z])([0-9][0-9])\-([0-9][0-9][0-9][0-9])([ve][0-9]+)(\..*)?$ $1/$1$2/$1$2-$3$4$5 [L,NC]

# http://www.aclweb.org/anthology/w/w10/W10-2410.pdf
# http://www.aclweb.org/anthology/W/w10/W10-2410.pdf - > http://www.aclweb.org//W/W10/W10-2410.pdf/W10-2410.pdf
# http://www.aclweb.org/anthology/w/W10/W10-2410.pdf - > http://www.aclweb.org//W/W10/W10-2410.pdf/W10/W10-2410.pdf
RewriteCond %{HTTP_HOST} ^anthology\.aclweb\.org$ [OR]
RewriteCond %{HTTP_HOST} ^www\.anthology\.aclweb\.org$
RewriteRule ^/?$ "http\:\/\/aclanthology\.info" [R=301,L]
```

**Comment by mbollmann** [comment]
> Here is the command I use for rsyncing to the Anthology after a build, in case it's useful:
> 
> ```
> $ rsync -azve ssh hugo/public/ hugo/data-export/ aclweb:/path/to/anthology
> ```

Getting a bit off-topic here, but:

1. `hugo/data-export` doesn't need to be copied, the contents are in `hugo/public` already.
2. We should ideally make sure that files are deleted if they're no longer there. For example, http://www.aclweb.org/anthology/people/h/hector-martinez-alonso1/ is a leftover from a previous build that should no longer be there. However, I guess we can't use the `--delete` flag since it would kill the PDFs? Not sure how it would work exactly...

**Comment by mjpost** [comment]
Good to know; I put that command here on a whim in hopes you'd correct it :)

Yes, agreed, I want to add `--delete` to the rsync. What's in the way is that I need a redirect rule to handle papers, which are currently a set of symlinked directories inside the directory I rsync into (e.g., `/path/to/anthology/A → /path/to/PDF/files`). Once I can figure out how to do those redirects, the Anthology directory can be a clean rsync. 

(This should be a simple rule, but the ones I've tried end up exposing the actual URL, and it makes me a bit nervous working with live wires, as it were).

**Comment by mjpost** [comment]
Back on topic: We've addressed your item #2, but I'm wary of adding complexity in the form of production / non-production, especially for a trivial issue like this. So I guess that means we stick with the longer name (which still isn't that complicated)?

**Comment by mjpost** [comment]
With commit 4eebf0abfd2bcc4aa7c232191639c73d32bc796a the redirection is implemented and is part of the repo, so I'm going to close this.
